arch/x86/real_prot_dance.S

```diff
@@ -25,388 +24,302 @@
-
-	/* save ss */
-	movw	%ss, %ax	
-	movw	%ax, seg_ss
-
-	/* save rsp */
-	mov	%rsp, %rax
-	mov	%rax, prot_rsp
-
-	/* This is a long indirect jump to long_to_prot32bit_code. From long mode to 32bit mode, indirect form is the only way*/
-	ljmpl	*(indir_jmp_long_2_real)
+    movw  %es, %ax
+    movw  %ax, _saved_es
+
+    movw  %gs, %ax
+    movw  %ax, _saved_gs
+
+    movw  %fs, %ax
+    movw  %ax, _saved_fs
+
+    movw  %ss, %ax
+    movw  %ax, _saved_ss
+
+    /* Save stack pointer */
+    mov   %rsp, _saved_sp
+
+    push  $__KERN_CS32
+    push  $.Lprot_mode
+    lretq
+END_FUNC(_long_to_real)
 
 .code32
-long_to_prot32bit_code:
+.align 16
+.Lprot_mode:
+    /* Load real mode accessible GDT and IDT */
+    lgdt  rmode_gdt_ptr
+    lidt  rmode_idt_ptr
+
+    /* Disable LME in EFER */
+    movl  $MSR_EFER, %ecx
+    rdmsr
+    and   $~EFER_LME, %eax
+    wrmsr
+
+    /* Disable paging to enter protected mode */
+    mov   %cr0, %eax
+    and   $~(X86_CR0_PG | X86_CR0_WP), %eax
+    mov   %eax, %cr0
+
+    /* Use protected mode data selector */
+    mov   $__KERN_DS32, %eax
+    mov   %eax, %ds
+
+    /* Use real mode accessible stack */
+    mov   rmode_stack, %esp
+
+    /* Use 16-bit protected mode segment selectors */
+    mov   $__KERN_DS16, %eax
+    mov   %eax, %ds
+    mov   %eax, %es
+    mov   %eax, %gs
+    mov   %eax, %fs
+    mov   %eax, %ss
+
+    /* Jump to real mode */
+    ljmp  $__KERN_CS16, $.Lreal_mode
 
-	movl	$__KERN_TRAMP_DS32, %eax
-	movl	%eax, %ds
+.code16
+.align 16
+.Lreal_mode:
+    /* Clear CR3 register */
+    xor   %eax, %eax
+    mov   %eax, %cr3
+
+    /* Disable protected mode and caching */
+    mov   %cr0, %eax
+    push  %eax
+    and   $~X86_CR0_PE, %eax
+    or    $(X86_CR0_CD | X86_CR0_NW), %eax
+    wbinvd
+    mov   %eax, %cr0
+
+    /* Set CS to first segment. It should cover the .text.rmode section */
+    ljmp  $0x0, $.Lset_rmode_cs
+.Lset_rmode_cs:
+
+    /* Set other segment selectors to the first segment.
+     * It should cover the .bss.rmode and first pages of
+     * .data.rmode sections
+     */
+    mov   %cs, %ax
+    mov   %ax, %ds
+    mov   %ax, %es
+    mov   %ax, %gs
+    mov   %ax, %fs
+    mov   %ax, %ss
+
+    /* CALL */
+
+    /* Restore protected mode and caching */
+    pop   %eax
+    mov   %eax, %cr0
+
+    /* Reload GDT and IDT pointers - just in case... */
+    lgdt  rmode_gdt_ptr
+    lidt  rmode_idt_ptr
+
+    /* Use protected mode data selector */
+    mov   $__KERN_DS32, %eax
+    mov   %eax, %ds
+    mov   %eax, %es
+    mov   %eax, %gs
+    mov   %eax, %fs
+    mov   %eax, %ss
+
+    ljmp  $__KERN_CS32, $.Lret_to_prot_mode
 
-	/* everything is saved, let's load up trampoline gdt*/
-	lgdt	trampgdt
+.code32
+.align 16
+.Lret_to_prot_mode:
+    mov   %cr4, %eax
+    or    $(X86_CR4_PAE | X86_CR4_PSE), %eax
+    mov   %eax, %cr4
+
+    /* Load temporarily real mode page tables.
+     * This is needed to enable long mode.
+     */
+    mov   $rm_l4_pt_entries, %eax
```

Comment:
I think this is good that we're establishing separate real mode page tables.
But is this needed?

Reply:
It is convenient. Having them always placed within real mode accessible range makes the dance simpler as we do not care about other PTs placement.